File: analyze-clinical-trials.js
#!/usr/bin/env node

/**
 * Analyze the structure and content of the clinical_trials table in Supabase
 */

const { createClient } = require('@supabase/supabase-js');

// Supabase connection details
const SUPABASE_URL = 'https://wwjorfctbizdhqkpduxt.supabase.co';
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Ind3am9yZmN0Yml6ZGhxa3BkdXh0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTI1ODMwMzksImV4cCI6MjA2ODE1OTAzOX0.PW5ZRSQsK9ij97v4xg7FLQPXEwmxtZC_Zlxdx3dJKnY';

// Create Supabase client
const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

async function analyzeClinicalTrials() {
  console.log('=== CLINICAL TRIALS TABLE ANALYSIS ===\n');

  try {
    // 1. Get total count of records
    const { count: totalRecords, error: countError } = await supabase
      .from('clinical_trials')
      .select('*', { count: 'exact', head: true });

    if (countError) throw countError;
    console.log(`Total records in clinical_trials table: ${totalRecords}\n`);

    // 2. Get sample records to analyze structure
    console.log('Fetching sample records...');
    const { data: sampleRecords, error: sampleError } = await supabase
      .from('clinical_trials')
      .select('*')
      .limit(5);

    if (sampleError) throw sampleError;

    if (sampleRecords && sampleRecords.length > 0) {
      console.log(`Successfully fetched ${sampleRecords.length} sample records\n`);

      // Analyze column population
      console.log('=== COLUMN POPULATION ANALYSIS ===');
      const columnStats = analyzeColumnPopulation(sampleRecords);
      
      for (const [col, stats] of Object.entries(columnStats)) {
        console.log(`\n${col}:`);
        console.log(`  - Populated: ${stats.populated}/${stats.total} (${stats.percentage.toFixed(1)}%)`);
        console.log(`  - Data type: ${stats.type}`);
        if (stats.sampleValues.length > 0) {
          console.log(`  - Sample values: ${JSON.stringify(stats.sampleValues.slice(0, 3))}`);
        }
      }

      // Analyze JSONB fields in detail
      console.log('\n\n=== JSONB FIELD STRUCTURE ANALYSIS ===');

      console.log('\n1. ELIGIBILITY_CRITERIA Structure:');
      analyzeJsonbField(sampleRecords, 'eligibility_criteria');

      console.log('\n2. LOCATIONS Structure:');
      analyzeJsonbField(sampleRecords, 'locations');

      console.log('\n3. CONTACT_INFO Structure:');
      analyzeJsonbField(sampleRecords, 'contact_info');

      // Display full sample records
      console.log('\n\n=== SAMPLE RECORDS (First 2) ===');
      sampleRecords.slice(0, 2).forEach((record, i) => {
        console.log(`\n--- Record ${i + 1} ---`);
        console.log(`ID: ${record.id || 'N/A'}`);
        console.log(`Trial ID: ${record.trial_id || 'N/A'}`);
        console.log(`Title: ${(record.title || 'N/A').substring(0, 100)}...`);
        console.log(`Status: ${record.status || 'N/A'}`);
        console.log(`Phase: ${record.phase || 'N/A'}`);
        console.log(`Sponsor: ${record.sponsor || 'N/A'}`);
        console.log(`Conditions: ${JSON.stringify(record.conditions || [])}`);
        console.log(`Interventions: ${JSON.stringify(record.interventions || [])}`);

        // Pretty print JSONB fields
        if (record.eligibility_criteria) {
          console.log('\nEligibility Criteria:');
          console.log(JSON.stringify(record.eligibility_criteria, null, 2));
        }

        if (record.locations && record.locations.length > 0) {
          console.log(`\nLocations (${record.locations.length} total):`);
          record.locations.slice(0, 2).forEach(loc => {
            console.log(JSON.stringify(loc, null, 2));
          });
        }

        if (record.contact_info) {
          console.log('\nContact Info:');
          console.log(JSON.stringify(record.contact_info, null, 2));
        }
      });
    }

    // 3. Analyze data patterns and issues
    console.log('\n\n=== DATA PATTERNS AND ISSUES ===');

    // Check for missing critical fields
    console.log('\nChecking for records with missing critical fields...');

    // Records without title
    const { count: noTitleCount } = await supabase
      .from('clinical_trials')
      .select('*', { count: 'exact', head: true })
      .is('title', null);
    console.log(`Records without title: ${noTitleCount || 0}`);

    // Records without conditions
    const { count: emptyConditionsCount } = await supabase
      .from('clinical_trials')
      .select('*', { count: 'exact', head: true })
      .eq('conditions', []);
    console.log(`Records with empty conditions array: ${emptyConditionsCount || 0}`);

    // Status distribution
    console.log('\nStatus distribution:');
    const { data: allRecords } = await supabase
      .from('clinical_trials')
      .select('status');

    if (allRecords) {
      const statusCounts = {};
      allRecords.forEach(record => {
        const status = record.status || 'unknown';
        statusCounts[status] = (statusCounts[status] || 0) + 1;
      });

      Object.entries(statusCounts)
        .sort((a, b) => b[1] - a[1])
        .forEach(([status, count]) => {
          const percentage = (count / allRecords.length * 100).toFixed(1);
          console.log(`  - ${status}: ${count} (${percentage}%)`);
        });
    }

    // Phase distribution
    console.log('\nPhase distribution:');
    const { data: phaseData } = await supabase
      .from('clinical_trials')
      .select('phase');

    if (phaseData) {
      const phaseCounts = {};
      phaseData.forEach(record => {
        const phase = record.phase || 'Not specified';
        phaseCounts[phase] = (phaseCounts[phase] || 0) + 1;
      });

      Object.entries(phaseCounts)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .forEach(([phase, count]) => {
          console.log(`  - ${phase}: ${count}`);
        });
    }

    // Date range analysis
    console.log('\nDate range analysis:');
    const { data: dateData } = await supabase
      .from('clinical_trials')
      .select('start_date, completion_date');

    if (dateData) {
      const startDates = dateData
        .map(r => r.start_date)
        .filter(d => d)
        .sort();
      const completionDates = dateData
        .map(r => r.completion_date)
        .filter(d => d)
        .sort();

      if (startDates.length > 0) {
        console.log(`  - Earliest start date: ${startDates[0]}`);
        console.log(`  - Latest start date: ${startDates[startDates.length - 1]}`);
      }

      if (completionDates.length > 0) {
        console.log(`  - Earliest completion date: ${completionDates[0]}`);
        console.log(`  - Latest completion date: ${completionDates[completionDates.length - 1]}`);
      }
    }

    // Most common conditions
    console.log('\n\nMost common conditions (top 10):');
    const { data: conditionData } = await supabase
      .from('clinical_trials')
      .select('conditions');

    if (conditionData) {
      const conditionCounts = {};
      conditionData.forEach(record => {
        if (record.conditions && Array.isArray(record.conditions)) {
          record.conditions.forEach(condition => {
            conditionCounts[condition] = (conditionCounts[condition] || 0) + 1;
          });
        }
      });

      Object.entries(conditionCounts)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .forEach(([condition, count]) => {
          console.log(`  - ${condition}: ${count}`);
        });
    }

    // Location analysis
    console.log('\n\nLocation countries (top 10):');
    const { data: locationData } = await supabase
      .from('clinical_trials')
      .select('locations');

    if (locationData) {
      const countryCounts = {};
      locationData.forEach(record => {
        if (record.locations && Array.isArray(record.locations)) {
          record.locations.forEach(location => {
            if (location.country) {
              countryCounts[location.country] = (countryCounts[location.country] || 0) + 1;
            }
          });
        }
      });

      Object.entries(countryCounts)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .forEach(([country, count]) => {
          console.log(`  - ${country}: ${count}`);
        });
    }

  } catch (error) {
    console.error('Error analyzing clinical trials:', error);
  }
}

function analyzeColumnPopulation(records) {
  if (!records || records.length === 0) return {};

  const allColumns = new Set();
  records.forEach(record => {
    Object.keys(record).forEach(key => allColumns.add(key));
  });

  const columnStats = {};

  Array.from(allColumns).sort().forEach(col => {
    let populated = 0;
    const sampleValues = [];
    const dataTypes = new Set();

    records.forEach(record => {
      const value = record[col];
      if (value !== null && value !== '' && (!Array.isArray(value) || value.length > 0)) {
        populated++;
        if (sampleValues.length < 3) {
          if (typeof value === 'object') {
            sampleValues.push(`${Array.isArray(value) ? 'Array' : 'Object'} with ${Array.isArray(value) ? value.length : Object.keys(value).length} items`);
          } else {
            const strValue = String(value);
            sampleValues.push(strValue.length > 50 ? strValue.substring(0, 50) + '...' : strValue);
          }
        }
        dataTypes.add(Array.isArray(value) ? 'array' : typeof value);
      }
    });

    columnStats[col] = {
      populated,
      total: records.length,
      percentage: (populated / records.length) * 100,
      type: Array.from(dataTypes).join(', ') || 'unknown',
      sampleValues
    };
  });

  return columnStats;
}

function analyzeJsonbField(records, fieldName) {
  const fieldValues = records
    .map(r => r[fieldName])
    .filter(v => v !== null && v !== undefined);

  if (fieldValues.length === 0) {
    console.log(`  No populated ${fieldName} fields found in sample`);
    return;
  }

  console.log(`  Found ${fieldValues.length} populated ${fieldName} fields`);

  // For array fields (like locations)
  if (Array.isArray(fieldValues[0])) {
    console.log(`  Type: Array of objects`);
    console.log(`  Array lengths: ${fieldValues.slice(0, 5).map(v => v.length).join(', ')}`);

    // Get sample structure from first non-empty array
    const nonEmptyArray = fieldValues.find(v => v.length > 0);
    if (nonEmptyArray && nonEmptyArray.length > 0) {
      console.log(`  Sample item structure:`);
      const sampleItem = nonEmptyArray[0];
      Object.entries(sampleItem).forEach(([key, val]) => {
        console.log(`    - ${key}: ${typeof val}`);
      });
    }
  }
  // For object fields
  else if (typeof fieldValues[0] === 'object') {
    console.log(`  Type: Object`);

    // Collect all keys across samples
    const allKeys = new Set();
    fieldValues.forEach(value => {
      if (typeof value === 'object' && value !== null) {
        Object.keys(value).forEach(key => allKeys.add(key));
      }
    });

    console.log(`  Keys found: ${Array.from(allKeys).sort().join(', ')}`);

    // Show sample structure
    console.log(`  Sample structure:`);
    const sample = fieldValues[0];
    Object.entries(sample).forEach(([key, val]) => {
      if (typeof val === 'string' || typeof val === 'number' || typeof val === 'boolean') {
        console.log(`    - ${key}: ${typeof val} = ${String(val).substring(0, 50)}`);
      } else {
        console.log(`    - ${key}: ${typeof val}`);
      }
    });
  }
}

// Run the analysis
analyzeClinicalTrials();

File: analyze_clinical_trials.py
#!/usr/bin/env python3
"""
Analyze the structure and content of the clinical_trials table in Supabase
"""

import os
import json
import supabase
from datetime import datetime
import pandas as pd
from typing import Dict, List, Any

# Supabase connection details
SUPABASE_URL = "https://wwjorfctbizdhqkpduxt.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Ind3am9yZmN0Yml6ZGhxa3BkdXh0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTI1ODMwMzksImV4cCI6MjA2ODE1OTAzOX0.PW5ZRSQsK9ij97v4xg7FLQPXEwmxtZC_Zlxdx3dJKnY"

def analyze_clinical_trials():
    """Analyze the clinical_trials table structure and content"""
    
    # Create Supabase client
    supabase_client = supabase.create_client(SUPABASE_URL, SUPABASE_KEY)
    
    print("=== CLINICAL TRIALS TABLE ANALYSIS ===\n")
    
    # 1. Get total count of records
    count_response = supabase_client.table('clinical_trials').select('*', count='exact').execute()
    total_records = count_response.count if hasattr(count_response, 'count') else 0
    print(f"Total records in clinical_trials table: {total_records}\n")
    
    # 2. Get sample records to analyze structure
    print("Fetching sample records...")
    sample_response = supabase_client.table('clinical_trials').select('*').limit(5).execute()
    
    if sample_response.data:
        print(f"Successfully fetched {len(sample_response.data)} sample records\n")
        
        # Analyze column population
        print("=== COLUMN POPULATION ANALYSIS ===")
        column_stats = analyze_column_population(sample_response.data)
        for col, stats in column_stats.items():
            print(f"\n{col}:")
            print(f"  - Populated: {stats['populated']}/{stats['total']} ({stats['percentage']:.1f}%)")
            print(f"  - Data type: {stats['type']}")
            if stats['sample_values']:
                print(f"  - Sample values: {stats['sample_values'][:3]}")
        
        # Analyze JSONB fields in detail
        print("\n\n=== JSONB FIELD STRUCTURE ANALYSIS ===")
        
        # Analyze eligibility_criteria
        print("\n1. ELIGIBILITY_CRITERIA Structure:")
        analyze_jsonb_field(sample_response.data, 'eligibility_criteria')
        
        # Analyze locations
        print("\n2. LOCATIONS Structure:")
        analyze_jsonb_field(sample_response.data, 'locations')
        
        # Analyze contact_info
        print("\n3. CONTACT_INFO Structure:")
        analyze_jsonb_field(sample_response.data, 'contact_info')
        
        # Display full sample records
        print("\n\n=== SAMPLE RECORDS (First 2) ===")
        for i, record in enumerate(sample_response.data[:2]):
            print(f"\n--- Record {i+1} ---")
            print(f"ID: {record.get('id', 'N/A')}")
            print(f"Trial ID: {record.get('trial_id', 'N/A')}")
            print(f"Title: {record.get('title', 'N/A')[:100]}...")
            print(f"Status: {record.get('status', 'N/A')}")
            print(f"Phase: {record.get('phase', 'N/A')}")
            print(f"Sponsor: {record.get('sponsor', 'N/A')}")
            print(f"Conditions: {record.get('conditions', [])}")
            print(f"Interventions: {record.get('interventions', [])}")
            
            # Pretty print JSONB fields
            if record.get('eligibility_criteria'):
                print("\nEligibility Criteria:")
                print(json.dumps(record['eligibility_criteria'], indent=2))
            
            if record.get('locations'):
                print(f"\nLocations ({len(record.get('locations', []))} total):")
                for loc in record.get('locations', [])[:2]:  # Show first 2 locations
                    print(json.dumps(loc, indent=2))
            
            if record.get('contact_info'):
                print("\nContact Info:")
                print(json.dumps(record['contact_info'], indent=2))
    
    # 3. Analyze data patterns and issues
    print("\n\n=== DATA PATTERNS AND ISSUES ===")
    
    # Check for missing critical fields
    print("\nChecking for records with missing critical fields...")
    
    # Records without title
    no_title = supabase_client.table('clinical_trials').select('id').is_('title', 'null').limit(5).execute()
    print(f"Records without title: {len(no_title.data) if no_title.data else 0}")
    
    # Records without conditions
    empty_conditions = supabase_client.table('clinical_trials').select('id').eq('conditions', []).limit(5).execute()
    print(f"Records with empty conditions array: {len(empty_conditions.data) if empty_conditions.data else 0}")
    
    # Status distribution
    print("\nStatus distribution:")
    status_query = supabase_client.table('clinical_trials').select('status').execute()
    if status_query.data:
        status_counts = {}
        for record in status_query.data:
            status = record.get('status', 'unknown')
            status_counts[status] = status_counts.get(status, 0) + 1
        
        for status, count in sorted(status_counts.items(), key=lambda x: x[1], reverse=True):
            print(f"  - {status}: {count} ({count/len(status_query.data)*100:.1f}%)")
    
    # Phase distribution
    print("\nPhase distribution:")
    phase_query = supabase_client.table('clinical_trials').select('phase').execute()
    if phase_query.data:
        phase_counts = {}
        for record in phase_query.data:
            phase = record.get('phase', 'Not specified')
            phase_counts[phase] = phase_counts.get(phase, 0) + 1
        
        for phase, count in sorted(phase_counts.items(), key=lambda x: x[1], reverse=True)[:10]:
            print(f"  - {phase}: {count}")
    
    # Date range analysis
    print("\nDate range analysis:")
    date_query = supabase_client.table('clinical_trials').select('start_date, completion_date').execute()
    if date_query.data:
        start_dates = [r['start_date'] for r in date_query.data if r.get('start_date')]
        completion_dates = [r['completion_date'] for r in date_query.data if r.get('completion_date')]
        
        if start_dates:
            print(f"  - Earliest start date: {min(start_dates)}")
            print(f"  - Latest start date: {max(start_dates)}")
        
        if completion_dates:
            print(f"  - Earliest completion date: {min(completion_dates)}")
            print(f"  - Latest completion date: {max(completion_dates)}")


def analyze_column_population(records: List[Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
    """Analyze which columns are populated and their data types"""
    if not records:
        return {}
    
    # Get all unique columns
    all_columns = set()
    for record in records:
        all_columns.update(record.keys())
    
    column_stats = {}
    
    for col in sorted(all_columns):
        populated = 0
        sample_values = []
        data_types = set()
        
        for record in records:
            value = record.get(col)
            if value is not None and value != '' and value != []:
                populated += 1
                if len(sample_values) < 3:
                    if isinstance(value, (dict, list)):
                        sample_values.append(f"{type(value).__name__} with {len(value)} items")
                    else:
                        sample_values.append(str(value)[:50] + '...' if len(str(value)) > 50 else str(value))
                data_types.add(type(value).__name__)
        
        column_stats[col] = {
            'populated': populated,
            'total': len(records),
            'percentage': (populated / len(records)) * 100,
            'type': ', '.join(data_types) if data_types else 'unknown',
            'sample_values': sample_values
        }
    
    return column_stats


def analyze_jsonb_field(records: List[Dict[str, Any]], field_name: str):
    """Analyze the structure of a JSONB field"""
    field_values = [r.get(field_name) for r in records if r.get(field_name)]
    
    if not field_values:
        print(f"  No populated {field_name} fields found in sample")
        return
    
    print(f"  Found {len(field_values)} populated {field_name} fields")
    
    # For list fields (like locations)
    if isinstance(field_values[0], list):
        print(f"  Type: Array of objects")
        print(f"  Array lengths: {[len(v) for v in field_values[:5]]}")
        
        # Get sample structure from first non-empty array
        for values in field_values:
            if values and len(values) > 0:
                print(f"  Sample item structure:")
                sample_item = values[0]
                for key, val in sample_item.items():
                    val_type = type(val).__name__
                    print(f"    - {key}: {val_type}")
                break
    
    # For object fields
    elif isinstance(field_values[0], dict):
        print(f"  Type: Object")
        
        # Collect all keys across samples
        all_keys = set()
        for value in field_values:
            if isinstance(value, dict):
                all_keys.update(value.keys())
        
        print(f"  Keys found: {sorted(all_keys)}")
        
        # Show sample structure
        print(f"  Sample structure:")
        sample = field_values[0]
        for key, val in sample.items():
            val_type = type(val).__name__
            if isinstance(val, (str, int, float, bool)):
                print(f"    - {key}: {val_type} = {str(val)[:50]}")
            else:
                print(f"    - {key}: {val_type}")


if __name__ == "__main__":
    try:
        analyze_clinical_trials()
    except Exception as e:
        print(f"Error analyzing clinical trials: {e}")
        import traceback
        traceback.print_exc()